File: Python/6.zig-zag-conversion.py
#
# @lc app=leetcode id=6 lang=python
#
# [6] ZigZag Conversion
#
# https://leetcode.com/problems/zigzag-conversion/description/
#
# algorithms
# Medium (33.66%)
# Likes:    1241
# Dislikes: 3744
# Total Accepted:    375.4K
# Total Submissions: 1.1M
# Testcase Example:  '"PAYPALISHIRING"\n3'
#
# The string "PAYPALISHIRING" is written in a zigzag pattern on a given number
# of rows like this: (you may want to display this pattern in a fixed font for
# better legibility)
# 
# 
# P   A   H   N
# A P L S I I G
# Y   I   R
# 
# 
# And then read line by line: "PAHNAPLSIIGYIR"
# 
# Write the code that will take a string and make this conversion given a
# number of rows:
# 
# 
# string convert(string s, int numRows);
# 
# Example 1:
# 
# 
# Input: s = "PAYPALISHIRING", numRows = 3
# Output: "PAHNAPLSIIGYIR"
# 
# 
# Example 2:
# 
# 
# Input: s = "PAYPALISHIRING", numRows = 4
# Output: "PINALSIGYAHRPI"
# Explanation:
# 
# P     I    N
# A   L S  I G
# Y A   H R
# P     I
# 
#

# @lc code=start
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows == 1 or numRows >= len(s):
            return s 
        
        res = [''] * numRows
        index, step = 0, 1
        
        for ch in s:
            res[index] += ch
            
            if index == 0:
                step = 1
            elif index == numRows-1:
                step = -1
            index += step
        
        return ''.join(res)
# @lc code=end

File: C++/6.zig-zag-conversion.cpp
/*
 * @lc app=leetcode.cn id=6 lang=cpp
 *
 *  [6] ZigZag Conversion
 */


#include <string>
#include<vector>
#include <iostream>
using namespace std;

// @lc code=start
class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows==1) return s;
        
        vector<string> tmp(numRows);
        
        int row=0, d=1;
        for (int i=0; i<s.size(); i++) {
            tmp[row].push_back(s[i]);
            if(row==0) 
                d = 1;
            else if (row==numRows-1)
                d = -1;
            row += d;  
        }

        string res;
        for (int i=0; i<numRows; i++) 
            res.append(tmp[i]);
        return res;
    }
};
// @lc code=end

int main() {
    Solution sol;
    string s = "PAYPALISHIRING";
    int numRows = 3;
    cout << sol.convert(s, numRows) <<endl;
    return 0;
}